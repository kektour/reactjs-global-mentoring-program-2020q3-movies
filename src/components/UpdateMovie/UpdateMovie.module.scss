@import 'src/styles/shared.scss';

.content {
  width: 675px;
  margin: auto;
  background-color: #232323;
  color: #fff;
  padding: 40px 30px 70px 85px;
  display: flex;
  flex-direction: column;
  border-radius: 5px;
  box-shadow: 9px 13px 24px -9px rgba(0, 0, 0, 0.7);
}

.contentHeader {
  text-transform: uppercase;
  font-size: 40px;
  margin: 0;
  margin-bottom: 40px;
  font-weight: normal;
}

.crossButton {
  @extend %cross-button;
  $size: 30px;
  width: $size;
  height: $size;
  margin-left: auto;
  margin-bottom: 24px;

  &::after {
    font-size: $size;
  }
}

.form {
  margin-right: 55px;
}

.inputRoot {
  margin-bottom: 30px;
}

.lastInputRoot {
  margin-bottom: 100px;
}

.buttonsContainer {
  display: flex;
  justify-content: flex-end;
}

.buttonReset {
  @extend %button;
  background-color: #232323;
  color: #f65261;
  border: 1px solid #f65261;
  margin-right: 15px;
}

.buttonSave {
  @extend %button;
  border: 1px solid #f65261;
}

.selectInput {
  height: 120px;
}