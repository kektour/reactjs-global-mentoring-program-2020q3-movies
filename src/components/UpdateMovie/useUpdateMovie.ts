import { AxiosError } from 'axios';
import { useState } from 'react';
import { Movie } from '../../models/movie';
import { updateMovie as uM } from '../../services/movies';

type UseUpdateMovie = {
  updateMovie: (m: Movie) => Promise<void>;
  error: Record<string, string> | null;
  clearError: () => void;
};

interface PostMovieErrorResponse {
  messages: Array<string>;
}

export const useUpdateMovie = (): UseUpdateMovie => {
  const [error, setError] = useState<Record<string, string> | null>(null);
  const updateMovie = async (m: Movie) => {
    try {
      uM(m);
    } catch (err) {
      const typedErr = err as AxiosError<PostMovieErrorResponse>;
      const errorMessages = getErrorMessages(typedErr.response!.data.messages);
      setError(errorMessages);
    }
  };
  const clearError = () => setError(null);

  return {
    updateMovie,
    error,
    clearError,
  };
};

const getErrorMessages = (arr: Array<string>): Record<string, string> =>
  arr.reduce((obj, v) => {
    let [fieldName] = v.split(' ');
    fieldName = fieldName.replace(/"/g, '');
    return {
      ...obj,
      [fieldName]: v,
    };
  }, {});
