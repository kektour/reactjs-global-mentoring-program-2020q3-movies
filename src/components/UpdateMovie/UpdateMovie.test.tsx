import React from 'react';
import ReactDOM from 'react-dom';

import { act, render, wait } from '@testing-library/react';
import { UpdateMovie } from './UpdateMovie';
import userEvent from '@testing-library/user-event';
import * as hooks from './useUpdateMovie';


describe('UpdateMovie', () => {
  const movie = {
    id: 337167,
    title: 'Fifty Shades Freed',
    tagline: "Don't miss the climax",
    vote_average: 6.1,
    vote_count: 1195,
    release_date: '2018-02-07',
    poster_path: 'https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg',
    overview:
      'Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.',
    budget: 55000000,
    revenue: 136906000,
    genres: ['Drama', 'Romance'],
    runtime: 106,
  };

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((e) => e) as any;

    const body = document.createElement('body');
    const div = document.createElement('div');
    const spyQuerySelector = jest.fn((selector: string) => {
      if (selector === '#root') {
        return div;
      } else if (selector === 'body') {
        return body;
      }
    });
    Object.defineProperty(global.document, 'querySelector', { value: spyQuerySelector });
  });

  it('Should render component correctly', () => {
    const component = render(<UpdateMovie movie={movie} open={true} onClose={() => null} />);
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('Should interact with the form', async () => {
    const { asFragment, container } = render(<UpdateMovie movie={movie} open={true} onClose={() => null} />);
    await act(async () => await userEvent.type(container.querySelector('input[name=title]') as Element, 'Title Foo'));
    await act(async () => await userEvent.type(container.querySelector('input[name=release_date]') as Element, '2019-09-04'));
    await act(async () => await userEvent.type(container.querySelector('input[name=poster_path]') as Element, 'Movie Preview Foo'));
    act(() => userEvent.selectOptions(container.querySelector('select[name=genres]') as Element, ['baz']));
    await act(async () => await userEvent.type(container.querySelector('input[name=overview]') as Element, 'Overview Foo'));
    await act(async () => await userEvent.type(container.querySelector('input[name=runtime]') as Element, '123'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should call API on submit form', async () => {
    const updateMovieMock = jest.fn().mockImplementation(() => Promise.resolve());
    const clearErrorMock = jest.fn();
    const handleCloseMock = jest.fn();
    const useUpdateMovieMock = jest.spyOn(hooks, 'useUpdateMovie').mockImplementation(() => ({ error: null, clearError: clearErrorMock, updateMovie: updateMovieMock }));

    const { container } = render(<UpdateMovie movie={movie} open={true} onClose={handleCloseMock} />);
    await act(async () => await userEvent.type(container.querySelector('input[name=title]') as Element, 'Title Foo'));
    await act(async () => await userEvent.type(container.querySelector('input[name=release_date]') as Element, '2019-09-04'));
    await act(async () => await userEvent.type(container.querySelector('input[name=poster_path]') as Element, 'Movie Preview Foo'));
    act(() => userEvent.selectOptions(container.querySelector('select[name=genres]') as Element, ['baz']));
    await act(async () => await userEvent.type(container.querySelector('input[name=overview]') as Element, 'Overview Foo'));
    await act(async () => await userEvent.type(container.querySelector('input[name=runtime]') as Element, '123'))
    act(() => userEvent.click(container.querySelector('button[type=submit]') as Element));

    await wait(() => {
      expect(updateMovieMock).toHaveBeenCalledWith({
        ...movie,
        title: 'Title Foo',
        release_date: '2019-09-04',
        poster_path: 'Movie Preview Foo',
        genres: ['baz'],
        overview: 'Overview Foo',
        runtime: '123',
      });
    });
    expect(handleCloseMock).toHaveBeenCalled();
    expect(clearErrorMock).toHaveBeenCalled();

    useUpdateMovieMock.mockRestore();
  });

  it('Should show errors', async () => {
    const useUpdateMovieMock = jest.spyOn(hooks, 'useUpdateMovie').mockImplementation(
      () =>
        ({
          error: {
            title: '"title" error title field',
            genres: '"genres" error genres field',
            overview: '"overview" error overview field',
          },
        } as any)
    );

    const { asFragment } = render(<UpdateMovie movie={movie} open={true} onClose={() => null} />);
    expect(asFragment()).toMatchSnapshot();

    useUpdateMovieMock.mockRestore();
  });
});
