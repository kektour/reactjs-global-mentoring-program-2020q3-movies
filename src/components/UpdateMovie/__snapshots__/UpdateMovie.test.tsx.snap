// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateMovie Should interact with the form 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Edit Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Id
            </p>
            <input
              class="input"
              name="id"
              readonly=""
              value="337167"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value="Title Foo"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2019-09-04"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="Movie Preview Foo"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value="Overview Foo"
            />
          </div>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="123"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UpdateMovie Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Edit Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Id
            </p>
            <input
              class="input"
              name="id"
              readonly=""
              value="337167"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value="Fifty Shades Freed"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2018-02-07"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value="Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins."
            />
          </div>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="106"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`UpdateMovie Should show errors 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="root header"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          <p
            class="root"
          >
            <span>
              netflix
            </span>
            roulette
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="content"
      >
        <button
          class="crossButton"
        />
        <h2
          class="contentHeader"
        >
          Edit Movie
        </h2>
        <form
          class="form"
        >
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Id
            </p>
            <input
              class="input"
              name="id"
              readonly=""
              value="337167"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Title
            </p>
            <input
              class="input"
              name="title"
              placeholder="Title here"
              value="Fifty Shades Freed"
            />
          </div>
          <p
            style="color: red;"
          >
            "title" error title field
          </p>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Release Date
            </p>
            <input
              class="input"
              name="release_date"
              placeholder="Release Date here"
              type="date"
              value="2018-02-07"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Movie Preview
            </p>
            <input
              class="input"
              name="poster_path"
              placeholder="Movie Preview here"
              value="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
            />
          </div>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Gende
            </p>
            <select
              class="root selectRoot selectInput"
              multiple=""
              name="genres"
            >
              <option
                disabled=""
                hidden=""
                value=""
              >
                Choose here
              </option>
              <option
                class="option selectOption"
                value="foo"
              >
                Foo
              </option>
              <option
                class="option selectOption"
                value="bar"
              >
                Bar
              </option>
              <option
                class="option selectOption"
                value="baz"
              >
                Baz
              </option>
            </select>
          </div>
          <p
            style="color: red;"
          >
            "genres" error genres field
          </p>
          <div
            class="root inputRoot"
          >
            <p
              class="title"
            >
              Overview
            </p>
            <input
              class="input"
              name="overview"
              placeholder="Overview here"
              value="Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins."
            />
          </div>
          <p
            style="color: red;"
          >
            "overview" error overview field
          </p>
          <div
            class="root inputRoot lastInputRoot"
          >
            <p
              class="title"
            >
              Runtime
            </p>
            <input
              class="input"
              name="runtime"
              placeholder="Runtime here"
              value="106"
            />
          </div>
          <div
            class="buttonsContainer"
          >
            <button
              class="buttonReset"
              type="reset"
            >
              Reset
            </button>
            <button
              class="buttonSave"
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      class="root footerRoot"
    >
      <p
        class="root logoRoot"
      >
        <span>
          netflix
        </span>
        roulette
      </p>
    </div>
  </div>
</DocumentFragment>
`;
